### PR TITLE
fix(saml): update dependencies and integration for ZITADEL V2

### PR DESCRIPTION
**Pull request by stebenz**

### CODE REVIEW

File: pkg/provider/metadata.go [hunk 1]
```diff
@@ -69,47 +77,55 @@ func (p *IdentityProviderConfig) getMetadata(
...
 				Location:  p.Endpoints.Artifact.URL,
-			}},
+			}},*/
 			SingleLogoutService: []md.EndpointType{
-				{
+				/*{
 					Binding:  SOAPBinding,
 					Location: p.Endpoints.SLOArtifact.URL,
-				},
+				},*/
```
Comment (livio-a) [inline pkg/provider/metadata.go:121]: please remove comments or code

File: pkg/provider/redirect.go [hunk 2]
```diff
@@ -81,14 +83,28 @@ func createRedirectSignature(
...
+	}
+
+	sig, err := signature.CreateRedirect(signingContext, buildRedirectQuery(string(respData), response.RelayState, idp.conf.SignatureAlgorithm, ""))
 	if err != nil {
 		return err
 	}
 
 	response.Signature = url.QueryEscape(base64.StdEncoding.EncodeToString(sig))
-	response.SigAlg = url.QueryEscape(base64.StdEncoding.EncodeToString([]byte(sigAlg)))
+	response.SigAlg = url.QueryEscape(base64.StdEncoding.EncodeToString([]byte(idp.conf.SignatureAlgorithm)))
```
Comment (livio-a) [inline pkg/provider/redirect.go:107]: why do you need to url encode after base64 encoding?

File: pkg/provider/xml/xml.go [hunk 3]
```diff
@@ -181,3 +176,29 @@ func DecodeLogoutRequest(encoding string, message string) (*samlp.LogoutRequestT
 
 	return req, nil
 }
+
+func DecodeResponse(encoding string, message string) (*samlp.ResponseType, error) {
+	/*reqBytes, err := base64.StdEncoding.DecodeString(message)
+	if err != nil {
+		return nil, err
+	}*/
```
Comment (livio-a) [inline pkg/provider/xml/xml.go:184]: please remove

File: pkg/provider/metadata.go [hunk 4]
```diff
@@ -1,21 +1,26 @@
...
 	"github.com/zitadel/saml/pkg/provider/xml/md"
 	"github.com/zitadel/saml/pkg/provider/xml/xenc"
 	"github.com/zitadel/saml/pkg/provider/xml/xml_dsig"
 	"net/http"
+	"time"
+)
+
+const (
+	DefaultValidUntil    time.Duration = 5 * time.Minute
+	DefaultCacheDuration string        = "PT300S"
```
Comment (livio-a) [inline pkg/provider/metadata.go:19]: PT300S?

File: pkg/provider/signature/certificates.go [hunk 5]
```diff
@@ -84,6 +84,33 @@ func GetSigningContextAndSigner(
...
+func GetSigner(
+	cert []byte,
+	key *rsa.PrivateKey,
+	signatureAlgorithm string,
+) (xmlsig.Signer, error) {
+	if signatureAlgorithm != dsig.RSASHA1SignatureMethod &&
+		signatureAlgorithm != dsig.RSASHA256SignatureMethod &&
+		signatureAlgorithm != dsig.RSASHA512SignatureMethod {
+		return nil, fmt.Errorf("invalid signing method %s", signatureAlgorithm)
+	}
```
Comment (livio-a) [inline pkg/provider/signature/certificates.go:96]: maybe create a separate function for the valid / accepted signature methods?

**Review by livio-a — CHANGES_REQUESTED** [review]
i went mostly through the changes... might need some review of the code itself some time

best might be to discuss it together

File: pkg/provider/identityprovider.go [hunk 6]
```diff
@@ -225,44 +215,51 @@ func (p *IdentityProvider) verifyRequestDestinationOfAttrQuery(request *samlp.At
...
+	if !ok || cert == nil || len(cert) == 0 {
+		return nil, nil, fmt.Errorf("failed to parse certificate")
+	}
+
+	keyWebKey := certAndKey.Key.Key.(jose.JSONWebKey)
+	if keyWebKey.Key == nil {
+		return nil, nil, fmt.Errorf("key is nil")
 	}
+	key, ok := keyWebKey.Key.(*rsa.PrivateKey)
+	if !ok || key == nil || reflect.DeepEqual(key, rsa.PrivateKey{}) {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply pkg/provider/identityprovider.go:244]: that was just something additional I thought would maybe make sense, as the logic for saml and the certificate generation are 2 different parts

File: pkg/provider/redirect.go (hunk 2 above)
Comment (stebenz) [reply pkg/provider/redirect.go:107]: to cover everything clearly, for example, if you refactor, by redirect-binding everything is in the URL

File: pkg/provider/metadata.go (hunk 4 above)
Comment (stebenz) [reply pkg/provider/metadata.go:19]: xs:duration type, https://www.data2type.de/xml-xslt-xslfo/xml-schema/datentypen-referenz/xs-duration, duration how long the receiver of the metadata is allowed to cache it

**Review by livio-a — APPROVED** [review]

**Pull request merged**